### ISSUE TITLE
Extensible loss and metric modules

### ISSUE THREAD

**Issue by tgaddair** [ludwig-ai/ludwig#1502]
Currently, users need to make changes to the underlying Ludwig internals in order to add a new loss or metric. Similar to what we did for encoders / decoders, it should be possible for users to add and register new losses and metrics using a simple decorator system like the following:

```
@loss("my_loss")
class MyLoss(nn.Module):
    def forward(self, preds: Tensor, target: Tensor) -> Tensor:
        ...

@metric("my_metric")
class MyMetric(Metric):
    def update(self, preds: Tensor, target: Tensor) -> None:
        ...

    def compute(self) -> Tensor:
        ...

    @abstractmethod
    def get_current_value(self, preds: Tensor, target: Tensor) -> Tensor:
        ...

@register_loss("my_loss", [NUMERICAL, CATEGORY])
@register_metric("my_metric", [NUMERICAL])
```

**Comment by w4nderlust** [comment]
The main complication here is that some losses / metrics may require predictions, others may require logins, or they may require probabilities, so the loss tensor association should be handled too.

**Comment by tgaddair** [comment]
Yes, I agree. Currently this is being done through a dictionary, but I think it would be easier if this were a property of the base class / mixin that defaults to logits. What do you think?

**Comment by justinxzhao** [comment]
I like the idea of having a registration mechanism.

Output feature classes already declare the loss function they use. Extending that to also include the specific tensor that the loss function operates on, as Travis suggests, seems reasonable.
